
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Pipelines &#8212; VATK 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Reporting alignment stats" href="reportingstats.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="pipelines">
<h1>Pipelines<a class="headerlink" href="#pipelines" title="Permalink to this headline">¶</a></h1>
<div class="section" id="hcmv-pipeline">
<h2>HCMV pipeline<a class="headerlink" href="#hcmv-pipeline" title="Permalink to this headline">¶</a></h2>
<p><strong>alignmentStats_ReCVR.sh</strong> is a bash script written to align multiple fastq files
to a reference sequence. The script expect 4 arguments.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>alignmentStats_ReCVR.sh &lt;input_dir&gt; &lt;ref.fa&gt; &lt;signature.fa&gt; &lt;lib_name&gt;
</pre></div>
</div>
<p><strong>Input</strong></p>
<p><strong>&lt;input_dir&gt;</strong> is a directory of paired fastq files. The fastq files need to be in a directory
and have the extension _R1_001.fastq and _R2_001.fastq (or _R1_001.fq and _R2_001.fq).</p>
<p><strong>&lt;ref.fa&gt;</strong> is the name of the reference fasta file (e.g., merlin.fa)</p>
<p><strong>&lt;signature.fa&gt;</strong> is the fasta file with the signatures of different HCMV strains.</p>
<p><strong>&lt;lib_name&gt;</strong> is the name of the reference library when building the bowtie2 index</p>
<div class="section" id="dependencies">
<h3>Dependencies<a class="headerlink" href="#dependencies" title="Permalink to this headline">¶</a></h3>
<p><a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">cutadapt</a></p>
<p><a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">trim_galore</a></p>
<p><a class="reference external" href="http://bowtie-bio.sourceforge.net/bowtie2/index.shtml">bowtie2</a></p>
<p><a class="reference external" href="https://sourceforge.net/projects/samtools/files/">samtools</a></p>
<p><a class="reference external" href="https://github.com/centre-for-virus-research/weeSAM/blob/master/legacy_versions/weeSAMv1.4">weeSAMv1.4</a></p>
<p><a class="reference external" href="https://www.gnu.org/software/gawk/">gawk</a></p>
<p><a class="reference external" href="https://github.com/centre-for-virus-research/VATK/blob/master/AssemblyPostProcessing/SamRemoveIndels.awk">SamRemoveIndels.awk</a> - hash-bang may need to be changed depending on your gawk installation</p>
<p><a class="reference external" href="https://github.com/centre-for-virus-research/VATK/blob/master/AssemblyPostProcessing/UniqSamPE.awk">UniqSamPE.awk</a> - hash-bang may need to be changed depending on your gawk installation</p>
<p><a class="reference external" href="https://github.com/centre-for-virus-research/VATK/tree/master/GenotypingTools">miRNA_Search</a></p>
</div>
<div class="section" id="pipeline">
<h3>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li>Step 1:</li>
</ul>
<p>Each fastq file in the folder is trimmed using <a class="reference external" href="https://github.com/FelixKrueger/TrimGalore">trim_galore</a> with the following settings
(–paired –length 21 –quality 10 –stringency 3).</p>
<ul class="simple">
<li>Step 2:</li>
</ul>
<p>The processed reads are subsequently aligned against the reference provided using bowtie2
allowing for a maximum fragment length of 1200 (-X 1200)</p>
<ul class="simple">
<li>Step 3:</li>
</ul>
<p>The assembly statistics are generated using <a class="reference external" href="https://github.com/centre-for-virus-research/weeSAM/blob/master/legacy_versions/weeSAMv1.4">weeSAMv1.4</a>.
A newer version of weeSAM is available <a class="reference external" href="https://github.com/centre-for-virus-research/weeSAM">here</a>
if you wish to have more comprehensive statistics.
A number of assembly statistics are also printed to the terminal.</p>
<ul class="simple">
<li>Step 4:</li>
</ul>
<p>The library diversity is also estimated, first using  <a class="reference external" href="https://github.com/centre-for-virus-research/VATK/blob/master/AssemblyPostProcessing/SamRemoveIndels.awk">SamRemoveIndels.awk</a>
and then with <a class="reference external" href="https://github.com/centre-for-virus-research/VATK/blob/master/AssemblyPostProcessing/UniqSamPE.awk">UniqSamPE.awk</a>.
These provide additional statistics which enable the calculation of the Ratio of total to unique coverage.</p>
<p>The diversity of genotypes in the sample is also estimated using <a class="reference external" href="https://github.com/centre-for-virus-research/VATK/tree/master/GenotypingTools">miRNA_Search</a>,
which used the signature motifs to determine the number of posible strains in the sample.</p>
<ul class="simple">
<li>Step 5:</li>
</ul>
<p>The final stats are printed to output.csv in the input directory.</p>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">VATK</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="evaluation.html">Evaluation of sequence data</a></li>
<li class="toctree-l1"><a class="reference internal" href="trimming.html">Trimming</a></li>
<li class="toctree-l1"><a class="reference internal" href="duplicate.html">Pre-assembly duplicate removal</a></li>
<li class="toctree-l1"><a class="reference internal" href="referenceselection.html">Reference selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment.html">Alignment/mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="postalignmentanalysis.html">Post-alignment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="reportingstats.html">Reporting alignment stats</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pipelines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#hcmv-pipeline">HCMV pipeline</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="reportingstats.html" title="previous chapter">Reporting alignment stats</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, CVR bioinformatics.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/pipelines.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>