# VATK
Virus Analysis Tool Kit


1. Evaluating your raw sequencing data
  * counting sequences: fastq and fastq.gz
  * quality assurance: FASTQC and samstats
  * Determining the quality and complexity of next-generation sequencing data without a reference genome 
2. trimming and de-multiplexing: Trimmomatic, Fastx_toolkit, fastx_trimmer, cutadapt, trim_galore, condetri, flexor (https://github.com/seqan/flexbar)
3. genotype detection, reference selection (Sreenuâ€™s tool), (de-novo assembly??)
4. alignment: bowtie2, bwa, tanoti
5. post alignment stats: wee_sam [https://github.com/centre-for-virus-research/weeSAM](https://github.com/centre-for-virus-research/weeSAM), count duplicates from SAM (Sreenu for Andrew)
6. post alignment analysis: haplotype window, remove primer reads, detecting chimeras???
7. consensus calling [https://github.com/vbsreenu/Sam2Consensus](https://github.com/vbsreenu/Sam2Consensus)
8. variant calling ([https://github.com/ZackBoyd123/valvs](https://github.com/ZackBoyd123/valvs))
