
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Trimming &#8212; VATK 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pre-assembly duplicate removal" href="duplicate.html" />
    <link rel="prev" title="Evaluation of sequence data" href="evaluation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="trimming">
<h1>Trimming<a class="headerlink" href="#trimming" title="Permalink to this headline">¶</a></h1>
<p>Sequencing data are usually provided as raw reads in the FASTQ format,
which usually require
pre-processing prior to further analysis.
Sequences often carry particular sequence oligonucleotides, These may have been
introduced by experimental design (e.g., sample barcodes), be specific to a sequencing
platform (e.g., adapter sequences) or be a real biological signal (e.g., spliced leader).
The most used pre-processing step
involves read trimming which:</p>
<ol class="arabic simple">
<li>Trims adaptors and primers from reads</li>
<li>Filters on base qualities, in different ways depending on the program</li>
</ol>
<p>It is important to trim and remove adaptors to avoid issues in downstream analyses such as
in transcriptomics <a class="reference internal" href="#williams2016" id="id1">[14]</a>, variant analysis
or de novo assembly.</p>
<p>There are many tools that have been developed for this purpose and an extensive
list of them are presented below. Each has its own advantage and some of them have already been benchmarked <a class="reference internal" href="#delfabbro2013" id="id2">[5]</a>.
Here tools for trimming and adaptor removal will be benchmarked
in the context of consensus calling and variant calling for viral genomes.
Duplicate removal and contaminant sequence filtering will be covered elsewhere.</p>
<div class="section" id="tools">
<h2>Tools<a class="headerlink" href="#tools" title="Permalink to this headline">¶</a></h2>
<table border="1" class="colwidths-given docutils" id="id17">
<caption><span class="caption-text">Trimming programs</span><a class="headerlink" href="#id17" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="34%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Name</th>
<th class="head">Website</th>
<th class="head">Publication</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Btrim</td>
<td><a class="reference external" href="http://graphics.med.yale.edu/trim/">http://graphics.med.yale.edu/trim/</a></td>
<td><a class="reference internal" href="#kong2011" id="id3">[8]</a></td>
</tr>
<tr class="row-odd"><td>ConDeTri</td>
<td><a class="reference external" href="https://github.com/linneas/condetri">https://github.com/linneas/condetri</a></td>
<td><a class="reference internal" href="#smeds2011" id="id4">[13]</a></td>
</tr>
<tr class="row-even"><td>Cutadapt</td>
<td><a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/">https://cutadapt.readthedocs.io/en/stable/</a></td>
<td><a class="reference internal" href="#ej200" id="id5">[10]</a></td>
</tr>
<tr class="row-odd"><td>FASTX</td>
<td><a class="reference external" href="http://hannonlab.cshl.edu/fastx_toolkit/">http://hannonlab.cshl.edu/fastx_toolkit/</a></td>
<td>N/A</td>
</tr>
<tr class="row-even"><td>Flexbar</td>
<td><a class="reference external" href="https://github.com/seqan/flexbar/wiki/Manual">https://github.com/seqan/flexbar/wiki/Manual</a></td>
<td><a class="reference internal" href="#dodt2012" id="id6">[7]</a></td>
</tr>
<tr class="row-odd"><td>PRINSEQ</td>
<td><a class="reference external" href="http://prinseq.sourceforge.net/index.html">http://prinseq.sourceforge.net/index.html</a></td>
<td><a class="reference internal" href="#schmieder2011" id="id7">[12]</a></td>
</tr>
<tr class="row-even"><td>SolexaQA + DynamicTrim</td>
<td><a class="reference external" href="http://solexaqa.sourceforge.net/">http://solexaqa.sourceforge.net/</a></td>
<td><a class="reference internal" href="#cox2010" id="id8">[4]</a></td>
</tr>
<tr class="row-odd"><td>Scythe</td>
<td><a class="reference external" href="https://github.com/vsbuffalo/scythe">https://github.com/vsbuffalo/scythe</a></td>
<td>N/A</td>
</tr>
<tr class="row-even"><td>Seqtk</td>
<td><a class="reference external" href="https://github.com/lh3/seqtk">https://github.com/lh3/seqtk</a></td>
<td>N/A</td>
</tr>
<tr class="row-odd"><td>Sickle</td>
<td><a class="reference external" href="https://github.com/najoshi/sickle">https://github.com/najoshi/sickle</a></td>
<td>N/A</td>
</tr>
<tr class="row-even"><td>TagDust2</td>
<td><a class="reference external" href="http://tagdust.sourceforge.net/">http://tagdust.sourceforge.net/</a></td>
<td><a class="reference internal" href="#lassmann2015" id="id9">[9]</a></td>
</tr>
<tr class="row-odd"><td>Trimmomatic</td>
<td><a class="reference external" href="http://www.usadellab.org/cms/index.php?page=trimmomatic">http://www.usadellab.org/cms/index.php?page=trimmomatic</a></td>
<td><a class="reference internal" href="#bolger2014" id="id10">[2]</a></td>
</tr>
<tr class="row-even"><td>Atropos</td>
<td><a class="reference external" href="https://github.com/jdidion/atropos">https://github.com/jdidion/atropos</a></td>
<td><a class="reference internal" href="#didion2017" id="id11">[6]</a></td>
</tr>
<tr class="row-odd"><td>fastp</td>
<td><a class="reference external" href="https://github.com/OpenGene/fastp">https://github.com/OpenGene/fastp</a></td>
<td><a class="reference internal" href="#chen2018" id="id12">[3]</a></td>
</tr>
<tr class="row-even"><td>FastqPuri</td>
<td><a class="reference external" href="https://github.com/jengelmann/FastqPuri">https://github.com/jengelmann/FastqPuri</a></td>
<td><a class="reference internal" href="#perez-rubio2019" id="id13">[11]</a></td>
</tr>
<tr class="row-odd"><td>ea-utils</td>
<td><a class="reference external" href="https://expressionanalysis.github.io/ea-utils/">https://expressionanalysis.github.io/ea-utils/</a></td>
<td><a class="reference internal" href="#aronesty2013" id="id14">[1]</a></td>
</tr>
<tr class="row-even"><td>TrimGalore</td>
<td><a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/">https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/</a></td>
<td>N/A</td>
</tr>
<tr class="row-odd"><td>BBDuk</td>
<td><a class="reference external" href="https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/">https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbduk-guide/</a></td>
<td>N/A</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="how-to-run">
<h2>How to run<a class="headerlink" href="#how-to-run" title="Permalink to this headline">¶</a></h2>
<div class="section" id="btrim">
<h3>Btrim:<a class="headerlink" href="#btrim" title="Permalink to this headline">¶</a></h3>
<p>Trims low quality at both or either ends of a fastq sequence, identify barcodes,
handle indels in adapters and barcodes. The adaptors are provided as a 6 column text-tab delimited
file where the user can specify which region in the sequence to search for adaptors in. See <a class="reference external" href="help/btrim.txt">btrim64</a>.
To trim Illumina adaptor AGATCGGAAGAGC with a window size of 10 and a cutoff for average
quality scores within the moving window for quality trimming of 25, where only 3’-end trimming is
carried out, keeping a minimal length sequence of 40:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>btrim64 -w <span class="m">10</span> -a <span class="m">25</span> -p illumina_adapter.txt -3 -P -o output.fastq -l <span class="m">40</span> -t &lt;<span class="o">(</span>gunzip -c path_to_your_fastq/*.gz <span class="o">)</span> -C -z
</pre></div>
</div>
<p>The disadvantage is that it does not deal directly with paired-end reads, so the individual
files need to be trimmed first and then the perl script paired_end_trim.pl is used
to create the files where both ends passed the trimming. Another two files (<a href="#id15"><span class="problematic" id="id16">*</span></a>.se) are
generated for the reads where only one end passed the trimming.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>btrim64 -a <span class="m">25</span> -q -t input_R1.fq -o output_R1.25 -s output_R1.sum25
btrim64 -a <span class="m">25</span> -q -t input_R1.fq -o output_R2.25 -s output_R2.sum25
paired_end_trim.pl output_R1.sum25 output_21.sum25 output_R1.25 output_R2.25
</pre></div>
</div>
<p>Works directly on .gz files. Provides total sequence count and length distribution.</p>
<p><strong>Default:</strong></p>
<ul class="simple">
<li>maximum number of errors in 5’-adaptor [default=3]</li>
<li>maximum number of errors in 3’-adaptor [default=4]</li>
<li>minimal insert size [default=25]</li>
<li>size of moving window for quality trimming [default=5]</li>
<li>cutoff for average quality scores within the moving window for quality trimming [default=15]</li>
</ul>
<p><strong>Disadvantages:</strong> Dealing with paired-end reads is a two step process.</p>
</div>
<div class="section" id="condetri">
<h3>ConDeTri:<a class="headerlink" href="#condetri" title="Permalink to this headline">¶</a></h3>
<p>Single-end and paired-end trimming independent from user
interaction, trim and remove reads with low quality scores. See  <a class="reference external" href="help/condetri.txt">condetri_v2.2.pl -h</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>condetri.pl -sc<span class="o">=</span><span class="m">33</span> -fastq1<span class="o">=</span>input_R1.fq -fastq2<span class="o">=</span>input_R2.fq -prefix<span class="o">=</span>output
</pre></div>
</div>
<p><strong>Default:</strong></p>
<ul class="simple">
<li>HQ = Hiqh quality threshold [25].</li>
<li>LQ = Low quality threshold [10].</li>
<li>Frac = Fraction of read that must exceed hq [0.8].</li>
<li>Minlen = Min allowed read length [50].</li>
<li>MH = When this number of sequential HQ bases is reached, the trimming stops [5].</li>
<li>ML = Max number of LQ bases allowed after a stretch of HQ bases from 3’-end [1].</li>
</ul>
<p><strong>Disadvantages:</strong> Seems very very slow.</p>
</div>
<div class="section" id="cutadapt">
<h3>Cutadapt:<a class="headerlink" href="#cutadapt" title="Permalink to this headline">¶</a></h3>
<p>See <a class="reference external" href="help/cutadapt.txt">cutadapt -help</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>--quality-base N      Assume that quality values in FASTQ are encoded as
                       ascii<span class="o">(</span>quality + N<span class="o">)</span>. This needs to be <span class="nb">set</span> to <span class="m">64</span> <span class="k">for</span>
                       some old Illumina FASTQ files. Default: <span class="m">33</span>
-m LEN<span class="o">[</span>:LEN2<span class="o">]</span>, --minimum-length LEN<span class="o">[</span>:LEN2<span class="o">]</span>
                       Discard reads shorter than LEN. Default: <span class="m">0</span>
  cutadapt -a ADAPT1 -A ADAPT2 <span class="o">[</span>options<span class="o">]</span> -o out1.fastq -p out2.fastq in1.fastq in2.fastq
</pre></div>
</div>
</div>
<div class="section" id="trim-galore">
<h3>Trim_galore<a class="headerlink" href="#trim-galore" title="Permalink to this headline">¶</a></h3>
<p>Trim_galore is a wrapper around cutadapt with a few additions.
See <a class="reference external" href="help/trimgalore.txt">trim_galore -h</a></p>
</div>
<div class="section" id="flexbar">
<h3>Flexbar:<a class="headerlink" href="#flexbar" title="Permalink to this headline">¶</a></h3>
<p>Can do color-space reads (454), maintains read pairing and separates barcode reads.</p>
<p>The program Flexbar preprocesses high-throughput sequencing data efficiently. It
demultiplexes barcoded runs and removes adapter sequences. Several adapter removal presets
for Illumina libraries are included. Flexbar computes exact overlap alignments using SIMD
and multicore parallelism. Moreover, trimming and filtering features are provided, e.g.
trimming of homopolymers at read ends. Flexbar increases read mapping rates and improves
genome as well as transcriptome assemblies. Unique molecular identifiers can be extracted
in a flexible way. The software supports data in fasta and fastq format from multiple
sequencing platforms. See <a class="reference external" href="help/flexbar.txt">flexbar -hh</a></p>
</div>
<div class="section" id="prinseq">
<h3>PRINSEQ:<a class="headerlink" href="#prinseq" title="Permalink to this headline">¶</a></h3>
<p>One of the big advantages of prinseq are the large number of different graphical outputs. See <a class="reference external" href="help/prinseq.txt">prinseq-lite.pl -help</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>perl prinseq-lite.pl -verbose -fastq test_1.fq -fastq2 test_2.fq -ns_max_n <span class="m">0</span> -out_good test_no_ns -out_bad test_with_ns
</pre></div>
</div>
<p>Singletons will automatically be filtered and stored in separate files (here test_no_ns_1_singletons.fastq and test_no_ns_2_singletons.fastq)
in addition to the files test_no_ns_1.fastq, test_no_ns_2.fastq, test_with_ns_1.fastq, and test_with_ns_2.fastq.</p>
<p><strong>Disadvantages:</strong> No direct way of dealing with paired end that are zipped, you will need to unzip on the fly:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>/prinseq &lt;your parameters&gt; &lt;<span class="o">(</span>gzip -d -c input.fq.gz<span class="o">)</span>
</pre></div>
</div>
<p>Or if you have paired-end, you need to unzip-trim (zip if you want to keep your reads compressed) and then match the pairs up again (<a class="reference external" href="https://github.com/sestaton/Pairfq">https://github.com/sestaton/Pairfq</a>)</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gzip -dc input_R1.fastq.gz <span class="p">|</span> prinseq-lite.pl -fastq stdin -out_bad null -out_good stdout <span class="se">\</span>
  -trim_qual_right <span class="m">20</span> -min_qual_mean <span class="m">20</span> -min_len <span class="m">35</span> -log prinseq.input_R1.log <span class="p">|</span> gzip &gt; input_R1.trim.gz
gzip -dc input_2.fastq.gz <span class="p">|</span> prinseq-lite.pl -fastq stdin -out_bad null -out_good stdout <span class="se">\</span>
  -trim_qual_right <span class="m">20</span> -min_qual_mean <span class="m">20</span> -min_len <span class="m">35</span> -log prinseq.input_R2.log <span class="p">|</span> gzip &gt; input_R2.trim.gz
pairfq makepairs -f input_R1.trim.gz -r input_R2.trim.gz -fp output_R1.trim.fastq-common.out.gz <span class="se">\</span>
  -rp output_R2.trim.fastq-common.out.gz -fs output_R1.trim.fastq-unique.out.gz -rs output_R2.trim.fastq-unique.out.gz --compress gzip
</pre></div>
</div>
</div>
<div class="section" id="bbduk">
<h3>BBDuk:<a class="headerlink" href="#bbduk" title="Permalink to this headline">¶</a></h3>
<p>One annoying thing about BBDuk.sh is that autocompletion of file names is reduced due to
the way arguments are specified for the bash script (e.g., -in1=input_R1.fq). BBDuk uses a kmer based
approach for doing the quality trimming. It also has a large number of other data filtering options.
See <a class="reference external" href="help/bbduk.txt">bbduk.sh -h</a></p>
</div>
<div class="section" id="trimmomatic">
<h3>Trimmomatic:<a class="headerlink" href="#trimmomatic" title="Permalink to this headline">¶</a></h3>
<p>A very highly cited program with a broad range of functionalities and reasonably straight forward to use.
It doesn’t have a command-line help but the manual PDF can be
found <a class="reference external" href="http://www.usadellab.org/cms/uploads/supplementary/Trimmomatic/TrimmomaticManual_V0.32.pdf">here</a>.</p>
</div>
</div>
<div class="section" id="benchmark">
<h2>Benchmark<a class="headerlink" href="#benchmark" title="Permalink to this headline">¶</a></h2>
<p>Current benchmarking pipeline available here <a class="reference external" href="https://github.com/centre-for-virus-research/VATK/blob/master/trimming/trimtest.snake">trimming/trimtest.snake</a></p>
<img alt="_images/dag.png" src="_images/dag.png" />
<p>Commands used for benchmarking:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>btrim64 -a <span class="m">25</span> -l <span class="m">50</span> -q -s tmp.r1.s -o output_R1 -t input_R1.fastq.gz
btrim64 -a <span class="m">25</span> -l <span class="m">50</span> -q -s tmp.r2.s -o output_R2 -t input_R2.fastq.gz
paired_end_trim.pl tmp.r1.s tmp.r2.s output_R1 output_R2
cat output_R1.se output_R2.se &gt; output_unpaired.fastq
mv output_R1.pe output_R1.fastq
mv output_R2.pe output_R2.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>condetri.pl -sc<span class="o">=</span><span class="m">33</span> -fastq1<span class="o">=</span>input_R1.fastq.gz -fastq2<span class="o">=</span>input_R2.fastq.gz -prefix<span class="o">=</span>output
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cutadapt -m <span class="m">50</span> -q <span class="m">25</span> -o output_R1.fastq.gz -p output_R2.fastq.gz --too-short-output<span class="o">=</span>tooshort.1.fastq.gz --too-short-paired-output<span class="o">=</span>tooshort.2.fastq.gz input_R1.fastq.gz input_R2.fastq.gz
cutadapt -m <span class="m">50</span> -q <span class="m">25</span> -o rescued.a.fastq.gz tooshort.1.fastq.gz
cutadapt -m <span class="m">50</span> -q <span class="m">25</span> -o rescued.b.fastq.gz tooshort.2.fastq.gz
cat  rescued.a.fastq.gz rescued.b.fastq.gz &gt; rescued.fastq.gz
gunzip -c rescued.fastq.gz &gt; output_unpaired.fastq
gunzip -c output_R1.fastq.gz &gt; output_R1.fastq
gunzip -c output_R2.fastq.gz &gt; output_R2.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>flexbar -r input_R1.fastq.gz -p input_R2.fastq.gz -qt <span class="m">25</span> -m <span class="m">50</span> -qf i1.8 --adapter-preset TruSeq -ap ON -s
mv flexbarOut_1.fastq output_R1.fastq
mv flexbarOut_2.fastq output_R2.fastq
cat flexbarOut_1_single.fastq flexbarOut_2_single.fastq &gt; output_unpaired.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gunzip -c input_R1.fastq.gz &gt; tmp_R1.fq
gunzip -c input_R2.fastq.gz &gt; tmp_R2.fq
prinseq-lite.pl -verbose -min_qual_mean <span class="m">25</span> -min_len <span class="m">50</span> -fastq tmp_R1.fq -fastq2 tmp_R2.fq -ns_max_n <span class="m">0</span> -out_good tmp_no_ns -out_bad null
cat tmp_no_ns_1_singletons.fastq tmp_no_ns_2_singletons.fastq &gt; output_unpaired.fastq
mv tmp_no_ns_1.fastq output_R1.fastq
mv tmp_no_ns_2.fastq output_R2.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>TrimmomaticPE -phred33 input_R1.fastq.gz input_R2.fastq.gz output_R1.fastq tmp1_unpaired.fq output_R1.fastq tmp2_unpaired.fq ILLUMINACLIP:TruSeq2-PE.fa:2:30:10 LEADING:3 TRAILING:3 SLIDINGWINDOW:5:25 MINLEN:50
cat tmp1_unpaired.fq tmp2_unpaired.fq &gt; output_unpaired.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>trim_galore --quality <span class="m">25</span> --length <span class="m">50</span> --retain_unpaired --stringency <span class="m">1</span> --dont_gzip --paired input_R1.fastq.gz input_R2.fastq.gz
cat input_R1_unpaired_1.fq input_R2_unpaired_2.fq &gt; output_unpaired.fastq
mv input_R1_001_val_1.fq output_R1.fastq
mv input_R2_001_val_2.fq output_R2.fastq
</pre></div>
</div>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>gunzip -c input_R1.fastq.gz &gt; tmp_R1.fq
gunzip -c input_R2.fastq.gz &gt; tmp_R2.fq
bbduk.sh <span class="nv">in1</span><span class="o">=</span>tmp_R1.fq <span class="nv">in2</span><span class="o">=</span>tmp_R2.fq <span class="nv">out1</span><span class="o">=</span>output_R1.fastq <span class="nv">out2</span><span class="o">=</span>output_R2.fastq <span class="nv">qtrim</span><span class="o">=</span>lr <span class="nv">trimq</span><span class="o">=</span><span class="m">25</span> <span class="nv">minlen</span><span class="o">=</span><span class="m">50</span> <span class="nv">rieb</span><span class="o">=</span><span class="nb">false</span> <span class="nv">outm1</span><span class="o">=</span>tmp1_unpaired.fq <span class="nv">outm2</span><span class="o">=</span>tmp2_unpaired.fq <span class="nv">overwrite</span><span class="o">=</span><span class="nb">true</span>
cat tmp1_unpaired.fq tmp2_unpaired.fq &gt; tmp_unpaired.fq
bbduk.sh <span class="nv">in</span><span class="o">=</span>tmp_unpaired.fq <span class="nv">out</span><span class="o">=</span>output_unpaired.fastq  <span class="nv">qtrim</span><span class="o">=</span>lr <span class="nv">trimq</span><span class="o">=</span><span class="m">25</span> <span class="nv">minlen</span><span class="o">=</span><span class="m">50</span> <span class="nv">overwrite</span><span class="o">=</span><span class="nb">true</span>
</pre></div>
</div>
<div class="section" id="time-comparison">
<h3>Time comparison<a class="headerlink" href="#time-comparison" title="Permalink to this headline">¶</a></h3>
<img alt="_images/time_plot.png" src="_images/time_plot.png" />
<p>btrim &lt; cutadapt &lt; trimgalore &lt; trimmomatic</p>
</div>
<div class="section" id="number-of-reads-remaining">
<h3>Number of reads remaining<a class="headerlink" href="#number-of-reads-remaining" title="Permalink to this headline">¶</a></h3>
<img alt="_images/count_plot.png" src="_images/count_plot.png" />
<p>More paired reads are removed for btrim, condetri and trimmomatic but a reasonable number
are rescued as single reads.</p>
</div>
<div class="section" id="number-of-bases-kept">
<h3>Number of bases kept<a class="headerlink" href="#number-of-bases-kept" title="Permalink to this headline">¶</a></h3>
<img alt="_images/base_plot.png" src="_images/base_plot.png" />
<p>The effect of trimming in terms of bases lost is not dramatically marked for btrim,
condetri, trimmomatic and bbduk.</p>
</div>
<div class="section" id="per-read-position-quality">
<h3>Per read position quality<a class="headerlink" href="#per-read-position-quality" title="Permalink to this headline">¶</a></h3>
<img alt="_images/qual_plot.png" src="_images/qual_plot.png" />
<p>The average quality is higher for trimmomatic and btrim. The decrease in quality over the
length of the read is not as steep. Condetri per base quality is slightly lower
but prinseq, trimgalore, cutadapt have even lower average quality and flexbar has the
lowest average quality.
Flexbar has very few single reads rescued, hence the distribution. Some further singleton
quality filtering is probably needed for cutadapt, trim_galore and flexbar.</p>
</div>
<div class="section" id="per-read-position-mismatch">
<h3>Per read position mismatch<a class="headerlink" href="#per-read-position-mismatch" title="Permalink to this headline">¶</a></h3>
<img alt="_images/mismatch_plot.png" src="_images/mismatch_plot.png" />
<p>Trimmomatic and btrim have very low levels of mismatch frequency with condetri having a
slightly higher mismatch frequency. Trimgalore, cutadapt and prinseq have higher levels of
mismatches along the length of the read. Flexbar has the worst level of mismatch frequency.</p>
<p>In the context of low-frequency variant calling, Trimmomatic and btrim are good options. BBDuk
also performs well, in this case it does not rescue any singletons but does not discard as
many reads it does this by trimming the length more agressively, which results in a higher
average quality to the right of the read. Condetri also performs reasonably well.</p>
<p>If there is a limited depth of coverage across the genome, then the poor trimming will
also affect the consensus calling so it is advisable to also use trimmomatic and btrim.</p>
<p>The tools that use a window or k-mer approach perform better in the metrics used here.</p>
</div>
</div>
<div class="section" id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p id="bibtex-bibliography-trimming-0"><table class="docutils citation" frame="void" id="aronesty2013" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id14">[1]</a></td><td>E&nbsp;Aronesty. Comparison of sequencing utility programs. open bioinforma j 7: 1–8. 2013.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="bolger2014" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id10">[2]</a></td><td>Anthony&nbsp;M Bolger, Marc Lohse, and Bjoern Usadel. Trimmomatic: a flexible trimmer for illumina sequence data. <em>Bioinformatics (Oxford, England)</em>, 30:2114–2120, August 2014. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btu170">doi:10.1093/bioinformatics/btu170</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="chen2018" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id12">[3]</a></td><td>Shifu Chen, Yanqing Zhou, Yaru Chen, and Jia Gu. Fastp: an ultra-fast all-in-one fastq preprocessor. <em>Bioinformatics (Oxford, England)</em>, 34:i884–i890, September 2018. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/bty560">doi:10.1093/bioinformatics/bty560</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="cox2010" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id8">[4]</a></td><td>Murray&nbsp;P Cox, Daniel&nbsp;A Peterson, and Patrick&nbsp;J Biggs. Solexaqa: at-a-glance quality assessment of illumina second-generation sequencing data. <em>BMC bioinformatics</em>, 11:485, September 2010. <a class="reference external" href="https://doi.org/10.1186/1471-2105-11-485">doi:10.1186/1471-2105-11-485</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="delfabbro2013" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[5]</a></td><td>Cristian Del&nbsp;Fabbro, Simone Scalabrin, Michele Morgante, and Federico&nbsp;M Giorgi. An extensive evaluation of read trimming effects on illumina ngs data analysis. <em>PloS one</em>, 8:e85024, 2013. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0085024">doi:10.1371/journal.pone.0085024</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="didion2017" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id11">[6]</a></td><td>John&nbsp;P Didion, Marcel Martin, and Francis&nbsp;S Collins. Atropos: specific, sensitive, and speedy trimming of sequencing reads. <em>PeerJ</em>, 5:e3720, 2017. <a class="reference external" href="https://doi.org/10.7717/peerj.3720">doi:10.7717/peerj.3720</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="dodt2012" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id6">[7]</a></td><td>Matthias Dodt, Johannes&nbsp;T Roehr, Rina Ahmed, and Christoph Dieterich. Flexbar-flexible barcode and adapter processing for next-generation sequencing platforms. <em>Biology</em>, 1:895–905, December 2012. <a class="reference external" href="https://doi.org/10.3390/biology1030895">doi:10.3390/biology1030895</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="kong2011" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id3">[8]</a></td><td>Yong Kong. Btrim: a fast, lightweight adapter and quality trimming program for next-generation sequencing technologies. <em>Genomics</em>, 98(2):152–153, 2011.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="lassmann2015" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id9">[9]</a></td><td>Timo Lassmann. Tagdust2: a generic method to extract reads from sequencing data. <em>BMC bioinformatics</em>, 16:24, January 2015. <a class="reference external" href="https://doi.org/10.1186/s12859-015-0454-y">doi:10.1186/s12859-015-0454-y</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="ej200" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id5">[10]</a></td><td>Marcel Martin. Cutadapt removes adapter sequences from high-throughput sequencing reads. <em>EMBnet.journal</em>, 17(1):10–12, 2011. URL: <a class="reference external" href="http://journal.embnet.org/index.php/embnetjournal/article/view/200">http://journal.embnet.org/index.php/embnetjournal/article/view/200</a>, <a class="reference external" href="https://doi.org/10.14806/ej.17.1.200">doi:10.14806/ej.17.1.200</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="perez-rubio2019" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id13">[11]</a></td><td>Paula Pérez-Rubio, Claudio Lottaz, and Julia&nbsp;C Engelmann. Fastqpuri: high-performance preprocessing of rna-seq data. <em>BMC bioinformatics</em>, 20:226, May 2019. <a class="reference external" href="https://doi.org/10.1186/s12859-019-2799-0">doi:10.1186/s12859-019-2799-0</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="schmieder2011" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id7">[12]</a></td><td>Robert Schmieder and Robert Edwards. Quality control and preprocessing of metagenomic datasets. <em>Bioinformatics (Oxford, England)</em>, 27:863–864, March 2011. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btr026">doi:10.1093/bioinformatics/btr026</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="smeds2011" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id4">[13]</a></td><td>Linnéa Smeds and Axel Künstner. Condetri–a content dependent read trimmer for illumina data. <em>PloS one</em>, 6:e26314, 2011. <a class="reference external" href="https://doi.org/10.1371/journal.pone.0026314">doi:10.1371/journal.pone.0026314</a>.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="williams2016" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[14]</a></td><td>Claire&nbsp;R Williams, Alyssa Baccarella, Jay&nbsp;Z Parrish, and Charles&nbsp;C Kim. Trimming of sequence reads alters rna-seq gene expression estimates. <em>BMC bioinformatics</em>, 17:103, February 2016. <a class="reference external" href="https://doi.org/10.1186/s12859-016-0956-2">doi:10.1186/s12859-016-0956-2</a>.</td></tr>
</tbody>
</table>
</p>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">VATK</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="evaluation.html">Evaluation of sequence data</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Trimming</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#tools">Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="#how-to-run">How to run</a></li>
<li class="toctree-l2"><a class="reference internal" href="#benchmark">Benchmark</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="duplicate.html">Pre-assembly duplicate removal</a></li>
<li class="toctree-l1"><a class="reference internal" href="referenceselection.html">Reference selection</a></li>
<li class="toctree-l1"><a class="reference internal" href="alignment.html">Alignment/mapping</a></li>
<li class="toctree-l1"><a class="reference internal" href="postalignmentanalysis.html">Post-alignment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="reportingstats.html">Reporting alignment stats</a></li>
<li class="toctree-l1"><a class="reference internal" href="pipelines.html">Pipelines</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="evaluation.html" title="previous chapter">Evaluation of sequence data</a></li>
      <li>Next: <a href="duplicate.html" title="next chapter">Pre-assembly duplicate removal</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, CVR bioinformatics.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.8.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/trimming.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>